libres/lib/res_util/subst_list.cpp

```diff
@@ -785,62 +785,113 @@ void subst_list_fprintf(const subst_list_type * subst_list , FILE * stream) {
 }
 
 
-/** Will loose tagging .... */
-int subst_list_add_from_string( subst_list_type * subst_list , const char * arg_string, bool append) {
-  int     error_count = 0;
-  if (arg_string != NULL) {
-    char ** key_value_list;
-    int     num_arg, iarg;
-
-    util_split_string(arg_string , "," , &num_arg , &key_value_list);
-    for (iarg = 0; iarg < num_arg; iarg++) {
-      if (strchr(key_value_list[iarg] , '=') == NULL)
-        //util_abort("%s: could not find \'=\' in argument string:%s \n",__func__ , key_value_list[iarg]);
-        /*
-          Could not find '=' in the argument string, this argument will
-          be ignored, and the error_count will be increased by one.
-        */
-        error_count += 1;
-      else {
-        char * key , * value;
-        char * tmp     = key_value_list[iarg];
-        int arg_length , value_length;
-        while (isspace(*tmp))  /* Skipping initial space */
-          tmp++;
-
-        arg_length = strcspn(tmp , " =");
-        key  = util_alloc_substring_copy(tmp , 0 , arg_length);
-        tmp += arg_length;
-        while ((*tmp == ' ') || (*tmp == '='))
-          tmp++;
-
-        value_length = strcspn(tmp , " ");
-        value = util_alloc_substring_copy( tmp , 0 , value_length);
-
-        /* Setting the argument */
+static int find_substring(const char *arg_string, const char *split_char)
+{
+    char pattern[4] = {'"', '\'', '\0', '\0'};
+    strcat(pattern, split_char);
+    int len = strcspn(arg_string, pattern);
+
+    while (strlen(arg_string) > len && (arg_string[len] == '"' || arg_string[len] == '\''))
+    {
+        const char delimiter = arg_string[len];
+
+        // Add the delimiter.
+        len++;
+
+        // The string must be long enough to accomdate a corresponding delimiter.
+        if (strlen(arg_string) <= len)
+            return -1;
+
+        // Find the corresponding delimiter.
+        const char *end = strchr(arg_string + len + 1, delimiter);
```

Comment:
The first argument to `strchr` should be a string right? I don't understand what the string
`arg_string + len + 1` becomes.